api/src/data_migration/transformation/subtask/transform_competition.py

```diff
@@ -0,0 +1,238 @@
+import logging
+
+from sqlalchemy import select
+
+import src.data_migration.transformation.transform_constants as transform_constants
+import src.data_migration.transformation.transform_util as transform_util
+from src.constants.lookup_constants import CompetitionOpenToApplicant, FormFamily
+from src.data_migration.transformation.subtask.abstract_transform_subtask import (
+    AbstractTransformSubTask,
+)
+from src.db.models.competition_models import Competition
+from src.db.models.opportunity_models import Opportunity
+from src.db.models.staging.competition import Tcompetition
+from src.db.models.staging.opportunity import TopportunityCfda
+
+logger = logging.getLogger(__name__)
+
+
+class TransformCompetition(AbstractTransformSubTask):
+    def transform_records(self) -> None:
+        logger.info("Processing competitions")
+
+        # Fetch all competitions that need to be processed
+        # Alongside that, grab the existing competition record
+        competitions: list[tuple[Tcompetition, Competition | None]] = self.fetch(
+            Tcompetition,
+            Competition,
+            [Tcompetition.comp_id == Competition.legacy_competition_id],
+        )
+
+        for source_competition, target_competition in competitions:
+            try:
+                opportunity_cfda = self.db_session.execute(
+                    select(TopportunityCfda).where(
+                        TopportunityCfda.opp_cfda_id == source_competition.opp_cfda_id
+                    )
+                ).scalar_one_or_none()
+
+                opportunity_id = None
+                opportunity_assistance_listing_id = None
+
+                if opportunity_cfda:
+                    opportunity_id = opportunity_cfda.opportunity_id
+                    opportunity_assistance_listing_id = opportunity_cfda.opp_cfda_id
+
+                    # Make sure the opportunity exists in our target table
+                    opportunity = self.db_session.execute(
+                        select(Opportunity).where(Opportunity.opportunity_id == opportunity_id)
+                    ).scalar_one_or_none()
+
+                    if not opportunity:
+                        logger.warning(
+                            "Competition references opportunity that doesn't exist in target schema",
+                            extra={
+                                "competition_id": source_competition.comp_id,
+                                "opportunity_id": opportunity_id,
+                            },
+                        )
```

Comment:
Let us not log a warning here - there is a non-problematic scenario where this can happen - if the opportunity was deleted, and we're processing the competition's delete, then this is actually expected (we would have recursively deleted the competition already - also why the if/elif/else is important in the other function).
```suggestion

```

Actually - we don't currently have a relationship between opportunity and competition (and opportunity should be aware of its competitions) - if we tried to delete the opportunity right now and it had a competition, we'd get a Postgres error - could we add that relationship with the cascade delete stuff?